### DISCUSSION TITLE
Refactoring the "streaming" terminology in "LockupDynamic"

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#708] [Ideas]
## Problem

It has become increasingly clear that the term `streams` is a misnomer for `LockupDynamic`. You can find a complete exposition of my rationale [here](https://github.com/sablier-labs/private-discussions/discussions/4), but, in short, the issue is two-fold:

1. The unlock-in-step curve is one of the payment options available on our official UI. Unlocking in steps is not streaming, so this is a case of cognitive dissonance on our part (I have been saying this [since we launched V2](https://github.com/sablier-labs/v2-interfaces/discussions/140#discussioncomment-6304651)).
2. People are using `LockupDynamic` for timelocks[^1][^2], i.e., "streams" in which the payout looks like this:

<img width="400" src="https://github.com/sablier-labs/v2-core/assets/8782666/355eee50-9be5-4fba-b2f7-9d44f1ffbb53"/>

## Solution

Rename the [`_streams` mapping](https://github.com/sablier-labs/v2-core/blob/6189bbc099d1d5fc50b62b93fd9a5fcfbadd7144/src/SablierV2LockupDynamic.sol#L63) to something else, as well as all semantically related references.

My suggestions:

- `payments`
- `plans`
- `agreements`
- `accords`
- `covenants`
- `schedules`
- `flows`

Although I will admit that I don't particularly like any one of these options, for instance, "payments" would be confusing because a settled stream may be composed of multiple withdrawals (the recipient may have withdrawn multiple times). And so you end up with a "payment of payments".

WDYT @andreivladbrg, @razgraf?

## Off-shoots

There are several consequences downstream, e.g., the `streamingBalanceOf` function would have to be replaced by `availableBalanceOf` (or something like that). But all these changes flow from the "streams" terminology, so let's deal with it first.

[^1]: https://github.com/sablier-labs/v2-core/issues/706
[^2]: https://twitter.com/ashleighschap/status/1712946120584781961

**Comment by PaulRBerg** [comment]
On second thoughts, "stream" might not be that bad terminology. Dictionary.com has this carve-out in the [definition for stream](https://www.dictionary.com/browse/stream):

<img width="400" alt="Screenshot 2023-10-15 at 1 59 58 PM" src="https://github.com/sablier-labs/v2-core/assets/8782666/31f77191-6d76-4fc3-81fb-cbe038f71e50">

I don't see why, in a future where the streaming meme grows in popularity, dictionary definitions could not be updated to include our kind of streams, i.e., something like "digital agreements for distributing financial assets over time".

**Comment by razgraf** [comment]
I'm not actually keen on changing this terminology. 

### Reason 1: What is streaming?

In tech, streams involve continuous distributions (what one might call live flows), but they're also based on clear intervals for said distribution. In video streaming, an asset is broken down into packages that are sent once every `x` milliseconds. In money streaming, at the moment, an asset is "realized" by the recipient once every `y` periods.

There's a cognitive load in assessing a stream split in longer segments as "streaming", I agree with that, but this doesn't mean it doesn't abide by the description above.

Therefore, depending on the definition you attach to the concept of streaming, the continuous aspect can simply define "a non-interrupted distribution of items" while the segments which make up that distribution (the milestones) can be on a spectrum from milliseconds to days, months, years.

> Unlocking in steps is not streaming

It can be. It isn't granular streaming, but in a definition that doesn't anchor to a set segment/period, it is still considered streaming.

### Reason 2: Brand identity

> People are using LockupDynamic for timelocks

If you mix and match, go to one extreme or the other, a stream/distribution can take many shapes. This has been proven by our LockupDynamic use-cases. Instead of changing the core concept, we could simply define sub-sets of streams. Instead of payroll, there is monthly streaming. Instead of step-unlocks there's quarterly streaming. Instead of complex shapes, there may be milestone streaming. Why change the term when we can adapt the meaning.

Secondly, why ditch a great word for a bland one when it become such an important part of our brand. It's a differentiator, a symbol. I've always been against brands losing their uniqueness, with a prime example being the [tweet-to-post](https://twitter.com/ubenjammin2/status/1687809950716694528) change. 

### Reason 3: People seem to get it

One important observation coming from these recent UNI-team-coin streams is that people seem to get it. They "created" this time-lock idea (which side note we'll probably add to the app in a bespoke manner soon) without external help. They molded a stream into a new shape. Sure, it's a small sample of users that confirms this reason (the same can be said for the OP) but it's enough to prompt us not to rush this decision of dropping/reducing the usage of "streaming" in our code/materials.

#### A separate suggestion

While I'd still use "streaming" as a base concept (thus keeping the `_streams` mapping as is), in the app we could simply split the `app.sablier.com/gallery` area of `Lockup Dynamic` templates into sections. Something like "regular" distributions vs. "sparse" unlocks (time-locks, steps, milestone based etc.). Of course, we need better terms for this, but if you're adamant on separating them from the bunch, we could simply do that from the UI.

**Reply by PaulRBerg** [reply]
Largely agree with what you said. Great points, and great idea to introduce a regular/sparse dichotomy at the UI level (might be worth opening a discussion so we don't forget about this!).

**Comment by PaulRBerg** [comment]
We talked about this on the call today (@razgraf, @andreivladbrg).

Arguments in favor of keeping the "streaming" terminology:

- People use Sablier for much more than just vesting, e.g., here's a gargantuan [600k USDC stream](https://app.sablier.com/?t=search&c=1&a=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48)
- No existing user or integrator has said that they are confused by this choice of terminology
- Replacing "streams" with "payments" would be dull (as @razgraf pointed out the example with Twitter's move from "tweets" to "posts" - the latter is boring AF, and everybody keeps saying "tweets")

We should monitor user feedback and come back to this if and when users give us signals that they are confused by this choice of words.

**Comment by PaulRBerg** [comment]
Related: https://github.com/sablier-labs/company-discussions/discussions/32

**Comment by PaulRBerg** [comment]
One more idea - replacing `streams` with `distributions`.

cc @sablier-labs/solidity.
